### PR TITLE
Generate precompile tests

### PR DESCRIPTION
**Pull request by ceyonur**
## Why this should be merged

Adds unit test generation to precompile generation tool

## How this works

Adds 2 new templates for contract and config tests.

## How this was tested

Generated hello world precompile in #564 

## How is this documented

Page at https://docs.avax.network/subnets/hello-world-precompile-tutorial needs to be updated.

### CODE REVIEW

File: accounts/abi/bind/precompilebind/precompile_contract_test_template.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	// This adds allowlist run tests to your custom tests
+	// and runs them all together.
+	// Even if you don't add any custom tests, keep this. This will still
+	// run the default allowlist tests.
+	allowlist.RunPrecompileWithAllowListTests(t, Module, state.NewTestStateDB, tests)
+	{{- else}}
+	// Run tests.
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			test.Run(t, module, NewTestStateDB(t *testing.T))
```
Comment (darioush) [inline accounts/abi/bind/precompilebind/precompile_contract_test_template.go:137]: `NewTestStateDB(t *testing.T)` does not seem correct here.

Is it possible to include a test that generates a precompile then compiles or calls gofmt on the resulting code? We could do this as a separate PR or an issue, however I think it will make it easier to edit the templates in the long run.

File: params/precompile_config_test.go [hunk 2]
```diff
@@ -64,7 +65,7 @@ func TestVerifyWithChainConfig(t *testing.T) {
 }
 
 func TestVerifyWithChainConfigAtNilTimestamp(t *testing.T) {
-	admins := []common.Address{{1}}
+	admins := []common.Address{allowlist.TestAdminAddr}
```
Comment (darioush) [inline params/precompile_config_test.go:68]: I think it would be nicer if params did not import allowlist

File: accounts/abi/bind/precompilebind/precompile_config_test_template.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+			Expected: false,
+		},
+		"different timestamp": {
+			Config:   NewConfig(big.NewInt(3),{{- if .Contract.AllowList}} admins, enableds{{- end}}),
+			Other:    NewConfig(big.NewInt(4),{{- if .Contract.AllowList}} admins, enableds{{- end}}),
+			Expected: false,
+		},
+		"same config": {
+			Config: NewConfig(big.NewInt(3),{{- if .Contract.AllowList}} admins, enableds{{- end}}),
+			Other: NewConfig(big.NewInt(3),{{- if .Contract.AllowList}} admins, enableds{{- end}}),
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply accounts/abi/bind/precompilebind/precompile_config_test_template.go]: correct, good catch

File: accounts/abi/bind/precompilebind/precompile_bind.go [hunk 4]
```diff
@@ -44,26 +44,56 @@ const (
 	readAllowListFuncKey = "readAllowList"
 )
 
+type BindedFiles struct {
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply accounts/abi/bind/precompilebind/precompile_bind.go:49]: added

File: accounts/abi/bind/precompilebind/precompile_contract_test_template.go (hunk 1 above)
Comment (ceyonur) [reply accounts/abi/bind/precompilebind/precompile_contract_test_template.go]: You mean like a unit test in the `precompile_bind_test`?

File: params/precompile_config_test.go (hunk 2 above)
Comment (ceyonur) [reply params/precompile_config_test.go]: yep good idea. actually `allowlist` is already in the import path for params tests as we import other precompile packages. but i agree, we should not import more precompile packages here.

**Review by aaronbuchwald — APPROVED** [review]

File: accounts/abi/bind/precompilebind/precompile_contract_template.go [hunk 5]
```diff
@@ -54,10 +54,10 @@ const (
 	// You should also increase gas costs of functions that read from AllowList storage.
 	{{- end}}
 	{{- range .Contract.Funcs}}
-	{{.Normalized.Name}}GasCost uint64 = 0 {{if not .Original.IsConstant | and $contract.AllowList}} + allowlist.ReadAllowListGasCost {{end}}	// SET A GAS COST HERE
+	{{.Normalized.Name}}GasCost uint64 = 1 /* SET A GAS COST HERE */ {{if not .Original.IsConstant | and $contract.AllowList}} + allowlist.ReadAllowListGasCost {{end}}
 	{{- end}}
 	{{- if .Contract.Fallback}}
-	{{.Contract.Type}}FallbackGasCost uint64 = 0 // SET A GAS COST LESS THAN 2300 HERE
+	{{.Contract.Type}}FallbackGasCost uint64 = 1 // SET A GAS COST LESS THAN 2300 HERE
```
[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply accounts/abi/bind/precompilebind/precompile_contract_template.go:60]: I just happened to run into this yesterday, so seems like a good change haha

File: accounts/abi/bind/precompilebind/precompile_contract_test_template.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	"github.com/ava-labs/subnet-evm/precompile/testutils"
+	"github.com/ava-labs/subnet-evm/vmerrs"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/stretchr/testify/require"
+)
+
+// TestRun tests the Run function of the precompile contract.
+// These tests are run against the precompile contract directly with
+// the given input and expected output. They're just a guide to
+// help you write your own tests. These test are for general cases like
```
Comment (darioush) [inline accounts/abi/bind/precompilebind/precompile_contract_test_template.go:29]:
```suggestion
// help you write your own tests. These tests are for general cases like
```

File: accounts/abi/bind/precompilebind/precompile_contract_test_template.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+// allowlist, readOnly behaviour and gas cost. You should write your own
```
Comment (darioush) [inline accounts/abi/bind/precompilebind/precompile_contract_test_template.go:30]:
```suggestion
// allowlist, readOnly behaviour, and gas cost. You should write your own
```

**Review by darioush — APPROVED** [review]

File: accounts/abi/bind/precompilebind/precompile_contract_test_template.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+		{{- range .Contract.Funcs}}
+		{{- $func := .}}
+		{{- if $contract.AllowList}}
+		{{- $roles := mkList "NoRole" "Enabled" "Admin"}}
+		{{- range $role := $roles}}
+		{{- $fail := and (not $func.Original.IsConstant) (eq $role "NoRole")}}
+		"calling {{decapitalise $func.Normalized.Name}} from {{$role}} should {{- if $fail}} fail {{- else}} succeed{{- end}}":  {
+			Caller:     allowlist.Test{{$role}}Addr,
+			BeforeHook: allowlist.SetDefaultRoles(Module.Address),
+			InputFn: func(t testing.TB) []byte {
```
Comment (aaronbuchwald) [inline accounts/abi/bind/precompilebind/precompile_contract_test_template.go:45]: can we gate the allow list specific tests to ensure that they are only generated (and the `BeforeHook` only specified) if the precompile is extending the allow list?
